numpyro/infer/hmc_gibbs.py

```diff
@@ -0,0 +1,108 @@
+# Copyright Contributors to the Pyro project.
+# SPDX-License-Identifier: Apache-2.0
+
+from collections import namedtuple
+import copy
+from functools import partial
+
+from jax import device_put, random, value_and_grad
+
+from numpyro.handlers import condition, seed, trace
+from numpyro.infer.mcmc import MCMCKernel
+from numpyro.infer.hmc import HMC, NUTS
+from numpyro.util import ravel_pytree
+
+
+HMCGibbs_State = namedtuple("HMCGibbs_State", "z, hmc_state, rng_key")
+"""
+ - **z** - a dict of the current latent values (both HMC and Gibbs sites)
+ - **hmc_state** - current hmc_state
+ - **rng_key** - random key for the current step
+"""
+
+
+def _wrap_model(model):
+    def fn(*args, **kwargs):
+        gibbs_values = kwargs.pop("_gibbs_sites", {})
+        with condition(data=gibbs_values):
+            model(*args, **kwargs)
+    return fn
+
+
+class HMCGibbs(MCMCKernel):
+    """
+    HMC-within-Gibbs. This inference algorithm allows the user to combine
+    general purpose gradient-based inference (HMC or NUTS) with custom
+    Gibbs samplers.
```

Comment:
nit: I guess this is already flexible enough for discrete and hmc_ecs cases. But can we mark the interface as EXPERIMENTAL? I would like to think how far we can compose this with the sequential proposal scheme in mixed hmc.